.github/workflows/backend.yml

```diff
@@ -44,3 +57,93 @@ jobs:
         push: ${{ startsWith(github.ref, 'refs/tags/') }}
         tags: parity/substrate-telemetry-backend:latest
         # add_git_labels: true
+
+  fmt:
+    name: Run rustfmt
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout sources
+        uses: actions/checkout@v2.3.4
+
+      - name: Install Rust stable toolchain
+        uses: actions-rs/toolchain@v1.0.7
+        with:
+          profile: minimal
+          toolchain: stable
+          override: true
+          components: clippy, rustfmt
+
+      - name: Rust Cache
+        uses: Swatinem/rust-cache@v1.3.0
+        with:
+          working-directory: backend
+
+      - name: Cargo fmt
+        run: cargo fmt --all -- --check
+
+  docs:
+    name: Check Documentation
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout sources
+        uses: actions/checkout@v2.3.4
+
+      - name: Install Rust stable toolchain
+        uses: actions-rs/toolchain@v1.0.7
+        with:
+          profile: minimal
+          toolchain: stable
+          override: true
+
+      - name: Rust Cache
+        uses: Swatinem/rust-cache@v1.3.0
+        with:
+          working-directory: backend
+
+      - name: Check internal documentation links
+        run: RUSTDOCFLAGS="--deny broken_intra_doc_links" cargo doc --verbose --workspace --no-deps --document-private-items
+
+  tests:
+    name: Run tests
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout sources
+        uses: actions/checkout@v2.3.4
+
+      - name: Install Rust stable toolchain
+        uses: actions-rs/toolchain@v1.0.7
+        with:
+          profile: minimal
+          toolchain: stable
+          override: true
+
+      - name: Rust Cache
+        uses: Swatinem/rust-cache@v1.3.0
+        with:
+          working-directory: backend
+
+      - name: Cargo test
+        run: cargo test --verbose --jobs 1
+
+  e2e:
+    name: Run potentially brittle E2E tests
+    continue-on-error: true
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout sources
+        uses: actions/checkout@v2.3.4
+
+      - name: Install Rust stable toolchain
+        uses: actions-rs/toolchain@v1.0.7
+        with:
+          profile: minimal
+          toolchain: stable
+          override: true
+
+      - name: Rust Cache
+        uses: Swatinem/rust-cache@v1.3.0
+        with:
+          working-directory: backend
+
+      - name: Cargo test
+        run: cargo test --verbose --jobs 1 --features e2e 'e2e'
```

Comment:
hrm, does this only work with one job at the time?

looks quite slow 15 mins :man_shrugging:

Reply:
Yeahh, I've noticed the e2e tests running somewhat slower than the rest. I think it's because I put them behind a feature flag, and more recompilation (possibly not cache friendly) happens as a result!

(`--jobs 1` runs 1 test at a time, but I don't think that's what makes this so slow)

Reply:
(I wonder whether it would be faster if I just mark them as ignored rather than use a feature flag, come to think of it.. I might try that first)

Reply:
(Yup, doing that made it much quicker; latest run was ~2 mins)